FEATURE: Change font methods to take a font size for most operations.

jarmstrong opened the pull request:
* ```load_font``` will currently provide a default size of 64.
* Calls to draw_text will load the font automatically for a given size if it has not already loaded.
* Fonts can be loaded early if desired by calling ```font_load_size```.
* ```font_has_size``` returns true if the font has been loaded with a given size.

macite commented:
Looks like it is including the commits I made earlier... @alexcu -- we had similar issues earlier, do you remember how to correct this?

jarmstrong commented:
Hold on, I think I know what happened, I will try closing and re-opening the request.

jarmstrong commented:
Alright looks like I've fixed it. I removed the last 3 commits that were on top of my history from you, and then did a ```git pull upstream develop --rebase``` and ```git push --force``` to my fork. Then the pull request updated itself. I think it was because I did a forced update of a commit I'd already pushed to my fork or something earlier.

macite commented on coresdk/src/coresdk/text.cpp:
```diff
@@ -18,18 +18,19 @@
+bool font_has_size(string name, int font_size)
+{
+    bool font_exists = has_font(name);
+    bool font_size_exists = _fonts[name]->_data.count(font_size) > 0;
```
You dont want to do this if the font doesn't exist.

jarmstrong replied:
Woops, I originally had it as a one liner so it would have not worked if the font didn't exist, slipped my mind.

jarmstrong replied:
Fixed in 81f1c3b

macite commented on coresdk/src/backend/text_driver.cpp:
```diff
@@ -131,7 +171,12 @@ void sk_draw_text(
+
+    // Check that the font size has been loaded and create if not
+    if (font->_data.count(font_size) == 0) {
+        font->_data[font_size] = TTF_OpenFont(font->filename.c_str(), font_size);
```
Add in checks to see if this worked. raise warning if it fails and return without drawing.

jarmstrong replied:
This one needs to stay I think, because the helper returns if the font for a size already exists, but we want the student to be able to call draw text with a valid font object but without having had loaded that font size?

macite replied:
I'd have the helper do the loading as well... if you ask it for a font at a size... it gives it to you.

macite commented on coresdk/src/backend/text_driver.cpp:
```diff
@@ -36,61 +37,99 @@ void sk_finalize_text()
+    if (VALID_PTR(font, FONT_PTR))
     {
-        cerr << "Error loading font " << SDL_GetError() << endl;
+        font->_data[font_size] = TTF_OpenFont(font->filename.c_str(), font_size);
```
Do we need to check if its not already loaded here? or is this done before?

jarmstrong replied:
Added a check.

macite commented on coresdk/src/backend/text_driver.cpp:
```diff
@@ -36,61 +37,99 @@ void sk_finalize_text()
-        return TTF_FontLineSkip(static_cast<TTF_Font *>(font->_data));
-    else
+    if (VALID_PTR(font, FONT_PTR)) {
+        return TTF_FontLineSkip(static_cast<TTF_Font *>(font->_data[font_size]));
```
Need to check that this size is loaded as well... maybe add in a helper function that returns the TTF_Font* from a font + size. Then it can load etc... all in the one place.

It could return nullptr if it fails, so that would be all you need to check in each place.

The helper could check that the font was valid as well etc..

macite replied:
Use this helper in the following code as well etc

jarmstrong replied:
Something like this?

```
TTF_Font* _get_font(sk_font_data* font, int font_size)
{
    if (VALID_PTR(font, FONT_PTR) and font->_data.count(font_size) > 0)
    {
        return static_cast<TTF_Font *>(font->_data[font_size]);
    }

    return nullptr;
}
```

I then use it like this...

```
int sk_text_line_skip(sk_font_data* font, int font_size)
{
    TTF_Font* ttf_font = _get_font(font, font_size);

    if (ttf_font)
    {
        return TTF_FontLineSkip(ttf_font);
    }
    else {
        return 0;
    }
}
```

Is that what you were thinking?

macite replied:
Yeah, basically. But you could also add in the code to load the font in the else path.

If you ask this for a font + size... and it delivers 😄

macite commented on coresdk/src/coresdk/text.cpp:
```diff
@@ -74,33 +75,31 @@ void free_all_fonts()
 }
 
-font load_font(string name, string filename, int size)
+font load_font(string name, string filename)
```
👍 looks good

macite commented on coresdk/src/coresdk/text.cpp:
```diff
@@ -133,17 +129,27 @@ font load_font(string name, string filename, int size)
 }
 
-void _print_strings(void *dest, font fnt, string str, rectangle rc, color fg_clr, color bg_clr, font_alignment flags)
+void font_load_size(string name, int font_size)
```
Provide this for `void font_load_size(font fnt, int font_size)` then call that here.

Makes this a one liner.

jarmstrong replied:
Done.

macite commented on coresdk/src/coresdk/text.h:
```diff
@@ -26,19 +26,23 @@ using namespace std;
 /// @constructor
 /// @csn initWithFontName:%s andSize:%s
-void set_font_style(font fnt, font_style style);
+void set_font_style(font fnt, int font_size, font_style style);
```
Should we just apply this to all?

jarmstrong replied:
Yeah that makes sense I will try to implement this soon.

macite commented on coresdk/src/coresdk/text.h:
```diff
@@ -26,19 +26,23 @@ using namespace std;
+void set_font_style(font fnt, int font_size, font_style style);
 
-font_style get_font_style(font fnt);
+font_style get_font_style(font fnt, int font_size);
```
Can remove size here as well -- just get from first.

macite commented on coresdk/src/coresdk/text.h:
```diff
@@ -26,19 +26,23 @@ using namespace std;
 
 bool has_font(string name);
 
+bool font_has_size(string name, int font_size);
```
Provide `font` versions of these as well.

macite commented:
Looking good... just need to think of how it could break :)

macite commented:
This is going to make text so much easier and more logic to work with! Thanks @jarmstrong

macite commented:
Looks good @jarmstrong. Just thought of one issue while reviewing related to style and sequence..

1. Load the font
1. Set style to italic
1. Draw font at different size -- wont be italic

Can you update this so that it applies the existing font style to new sizes of the font?

jarmstrong commented:
Ah, I just realized I may have made a mistake. The coresdk currently applies font styles to each font size, but I that should really be the backends job right?

macite commented:
That looks fine... I try to keep as much in the coresdk as I can then if we wanted to change the backend it shouldn't be too much work.
